ietf/doc/storage_backends.py

```diff
@@ -0,0 +1,137 @@
+
+    def __init__(self, **settings):
+        self.in_flight_custom_metadata: Dict[str, Dict[str, str]] = {}
+        return super().__init__(**settings)
+
+    def store_file(
+        self,
+        kind: str,
+        name: str,
+        file: Union[File, BufferedReader],
+        allow_overwrite: bool = False,
+        doc_name: Optional[str] = None,
+        doc_rev: Optional[str] = None,
+    ):
+        is_new = not self.exists_in_storage(kind, name)
+        # debug.show('f"Asked to store {name} in {kind}: is_new={is_new}, allow_overwrite={allow_overwrite}"')
+        if not allow_overwrite and not is_new:
+            log(f"Failed to save {kind}:{name} - name already exists in store")
+            debug.show('f"Failed to save {kind}:{name} - name already exists in store"')
+            debug.traceback()
+            raise Exception("Not ignoring overwrite attempts while testing")
+        else:
+            try:
+                new_name = self.save(name, file)
+                now = timezone.now()
+                existing_record = StoredObject.objects.filter(store=kind, name=name)
+                if existing_record.exists():
+                    # Note this is updating a queryset which is guaranteed by constraints to have one object
+                    existing_record.update(
+                        sha384=self.in_flight_custom_metadata[name]["sha384"],
+                        len=int(self.in_flight_custom_metadata[name]["len"]),
+                        modified=now,
+                    )
+                else:
+                    StoredObject.objects.create(
+                        store=kind,
+                        name=name,
+                        sha384=self.in_flight_custom_metadata[name]["sha384"],
+                        len=int(self.in_flight_custom_metadata[name]["len"]),
+                        store_created=now,
+                        created=now,
+                        modified=now,
+                        doc_name=doc_name,
+                        doc_rev=doc_rev,
+                    )
+                if new_name != name:
+                    complaint = f"Error encountered saving '{name}' - results stored in '{new_name}' instead."
+                    log(complaint)
+                    debug.show("complaint")
+                    # Note that we are otherwise ignoring this condition - it should become an error later.
+            except Exception as e:
+                # Log and then swallow the exception while we're learning.
+                # Don't let failure pass so quietly when these are the autoritative bits.
+                log(f"Failed to save {kind}:{name}", e)
+                raise e
+                debug.show("type(e)")
+                debug.show("e")
+                debug.traceback()
+            finally:
+                del self.in_flight_custom_metadata[name]
+        return None
+
+    def exists_in_storage(self, kind: str, name: str) -> bool:
```

Comment:
again, why not override?

Reply:
again, parallel structure. Also I think we want to (for the short term) keep the parent `.exists()` bare outside this overridden class just for continued learning.